# Entrega 2: Perfilado, Diccionario y Limpieza de Datos

Este notebook realiza la limpieza, formateo y transformación del dataset `Sumaria-2024.csv` de ENAHO, siguiendo las reglas establecidas en el diccionario de datos del proyecto.

## Unidad de análisis y granularidad

Unidad de análisis: hogar encuestado en ENAHO Sumaria 2024.

Granularidad: cada fila representa un hogar observado en un mes de levantamiento, con ubicación geográfica, condición socioeconómica, ingresos, gastos, composición del hogar y factor de expansión.

El dataset se usará como una tabla analítica a nivel de hogar para conectar posteriormente con Tableau.

In [2]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# 1. Carga de Datos
df_original = pd.read_csv('../Data/original/Sumaria-2024.csv', encoding='latin1')
print(f'Dimensiones originales: {df_original.shape}')

Dimensiones originales: (33691, 163)


## Perfilado inicial del dataset original

En esta sección se revisan dimensiones, tipos de variables, nulos, duplicados y cardinalidad inicial para identificar problemas de calidad antes de la limpieza.

In [3]:
print("Filas y columnas del dataset original:", df_original.shape)
print("Duplicados en dataset original:", df_original.duplicated().sum())

perfil_original = pd.DataFrame({
    "variable": df_original.columns,
    "tipo_original": df_original.dtypes.astype(str).values,
    "nulos": df_original.isnull().sum().values,
    "pct_nulos": (df_original.isnull().mean() * 100).round(2).values,
    "cardinalidad": df_original.nunique(dropna=True).values
})

perfil_original

Filas y columnas del dataset original: (33691, 163)
Duplicados en dataset original: 0


,variable,tipo_original,nulos,pct_nulos,cardinalidad
0,AÑO,int64,0,0.0,1
1,MES,int64,0,0.0,12
2,CONGLOME,int64,0,0.0,5359
3,VIVIENDA,int64,0,0.0,541
4,HOGAR,int64,0,0.0,15
...,...,...,...,...,...
158,FACTOR07,float64,0,0.0,968
159,LINEAV,float64,0,0.0,33488
160,POBREZAV,int64,0,0.0,4
161,NCONGLOME,int64,0,0.0,5323


## 2. Selección de Variables Relevantes
Nos quedamos estrictamente con las 18 variables definidas en la propuesta.

Estas variables fueron seleccionadas porque permiten responder la pregunta analítica del proyecto: analizar brechas socioeconómicas, diferencias territoriales, composición del gasto y variaciones temporales en hogares peruanos.

In [4]:
columnas_seleccionadas = [
    'MES', 'UBIGEO', 'DOMINIO', 'ESTRATO', 'POBREZA', 'POBREZAV',
    'INGHOG2D', 'GASHOG2D', 'MIEPERHO',
    'GRU11HD', 'GRU21HD', 'GRU31HD', 'GRU41HD', 'GRU51HD', 'GRU61HD', 'GRU71HD', 'GRU81HD',
    'FACTOR07'
]

df = df_original[columnas_seleccionadas].copy()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 33691 entries, 0 to 33690
Data columns (total 18 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   MES       33691 non-null  int64  
 1   UBIGEO    33691 non-null  int64  
 2   DOMINIO   33691 non-null  int64  
 3   ESTRATO   33691 non-null  int64  
 4   POBREZA   33691 non-null  int64  
 5   POBREZAV  33691 non-null  int64  
 6   INGHOG2D  33691 non-null  float64
 7   GASHOG2D  33691 non-null  float64
 8   MIEPERHO  33691 non-null  int64  
 9   GRU11HD   33691 non-null  float64
 10  GRU21HD   33691 non-null  float64
 11  GRU31HD   33691 non-null  float64
 12  GRU41HD   33691 non-null  float64
 13  GRU51HD   33691 non-null  float64
 14  GRU61HD   33691 non-null  float64
 15  GRU71HD   33691 non-null  float64
 16  GRU81HD   33691 non-null  float64
 17  FACTOR07  33691 non-null  float64
dtypes: float64(11), int64(7)
memory usage: 4.6 MB


## 2.1 Perfilado de variables seleccionadas

Se valida el subconjunto de variables que será usado en la limpieza inicial y en Tableau. Esta tabla permite revisar tipos, nulos, porcentaje de nulos y cardinalidad por campo.

In [5]:
print("Filas y columnas del dataset de trabajo:", df.shape)
print("Duplicados en variables seleccionadas:", df.duplicated().sum())

perfil_trabajo = pd.DataFrame({
    "variable": df.columns,
    "tipo": df.dtypes.astype(str).values,
    "nulos": df.isnull().sum().values,
    "pct_nulos": (df.isnull().mean() * 100).round(2).values,
    "cardinalidad": df.nunique(dropna=True).values
})

perfil_trabajo

Filas y columnas del dataset de trabajo: (33691, 18)
Duplicados en variables seleccionadas: 0


,variable,tipo,nulos,pct_nulos,cardinalidad
0,MES,int64,0,0.0,12
1,UBIGEO,int64,0,0.0,1271
2,DOMINIO,int64,0,0.0,8
3,ESTRATO,int64,0,0.0,8
4,POBREZA,int64,0,0.0,3
5,POBREZAV,int64,0,0.0,4
6,INGHOG2D,float64,0,0.0,33644
7,GASHOG2D,float64,0,0.0,33675
8,MIEPERHO,int64,0,0.0,17
9,GRU11HD,float64,0,0.0,33330


## 2.2 Campos críticos

| Campo | Rol analítico | Justificación |
|---|---|---|
| MES | Temporal | Permite analizar variaciones mensuales durante 2024. |
| UBIGEO | Geográfico | Permite ubicar territorialmente los hogares. |
| DOMINIO | Segmentación territorial | Permite comparar costa, sierra, selva y Lima Metropolitana. |
| ESTRATO | Segmentación poblacional | Permite comparar hogares según estrato. |
| POBREZA | Condición socioeconómica | Permite distinguir hogares pobres, pobres extremos y no pobres. |
| POBREZAV | Vulnerabilidad | Permite identificar hogares vulnerables o no vulnerables. |
| INGHOG2D | Métrica principal | Representa el ingreso del hogar. |
| GASHOG2D | Métrica principal | Representa el gasto del hogar. |
| MIEPERHO | Normalización | Permite considerar el tamaño del hogar. |
| GRU11HD-GRU81HD | Estructura del gasto | Permite analizar composición del consumo por rubro. |
| FACTOR07 | Ponderación | Permite aplicar el factor de expansión de la encuesta. |

## 3. Limpieza y Transformación (Mapping)

Reglas aplicadas:

1. **UBIGEO:** formato string de 6 dígitos con zero-padding. Se conserva como identificador geográfico, no como variable numérica.
2. **MES:** formato string de 2 dígitos con zero-padding para conservar consistencia temporal.
3. **Categóricas:** `DOMINIO`, `ESTRATO`, `POBREZA` y `POBREZAV` se homologan mediante diccionarios de etiquetas.
4. **Continuas:** ingresos, gastos, tamaño del hogar, grupos de gasto y factor de expansión se mantienen como variables numéricas.
5. **Duplicados y nulos:** se revisan antes y después de la limpieza.
6. **Outliers:** se imputan con la mediana del grupo de pobreza (ver Sección 4).
7. **Variables derivadas:** se construyen `BRECHA_HOG`, per cápita, flag de déficit y variables de soporte para visualización (ver Secciones 4.2 y 4.3).
8. **Exportación:** se genera un CSV limpio preliminar compatible con Tableau.

In [6]:
# Formateo de Strings (UBIGEO y MES)
df['UBIGEO'] = df['UBIGEO'].astype(str).str.replace(r'\.0$', '', regex=True).str.zfill(6)
df['MES'] = df['MES'].astype(str).str.replace(r'\.0$', '', regex=True).str.zfill(2)

# Mapeo DOMINIO
dic_dominio = {
    1: 'Costa Norte', 2: 'Costa Centro', 3: 'Costa Sur',
    4: 'Sierra Norte', 5: 'Sierra Centro', 6: 'Sierra Sur',
    7: 'Selva', 8: 'Lima Metropolitana'
}
df['DOMINIO'] = df['DOMINIO'].map(dic_dominio)

# Mapeo ESTRATO
dic_estrato = {
    1: 'De 500,000 a más habitantes',
    2: 'De 100,000 a 499,999 habitantes',
    3: 'De 50,000 a 99,999 habitantes',
    4: 'De 20,000 a 49,999 habitantes',
    5: 'De 2,000 a 19,999 habitantes',
    6: 'De 500 a 1,999 habitantes',
    7: 'Área de Empadronamiento Rural (AER) Compuesto',
    8: 'Área de Empadronamiento Rural (AER) Simple'
}
df['ESTRATO'] = df['ESTRATO'].map(dic_estrato)

# Mapeo POBREZA
dic_pobreza = {
    1: 'Pobre Extremo',
    2: 'Pobre No Extremo',
    3: 'No Pobre'
}
df['POBREZA'] = df['POBREZA'].map(dic_pobreza)

# Mapeo POBREZAV
dic_pobrezav = {
    1: 'Pobre Extremo',
    2: 'Pobre No Extremo',
    3: 'Vulnerable No Pobre',
    4: 'No Vulnerable'
}
df['POBREZAV'] = df['POBREZAV'].map(dic_pobrezav)

# Asegurar numéricas continuas
num_cols = ['INGHOG2D', 'GASHOG2D', 'MIEPERHO', 'FACTOR07'] + [f'GRU{i}1HD' for i in range(1, 9)]
for col in num_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

df.head()

,MES,UBIGEO,DOMINIO,ESTRATO,POBREZA,POBREZAV,INGHOG2D,GASHOG2D,MIEPERHO,GRU11HD,GRU21HD,GRU31HD,GRU41HD,GRU51HD,GRU61HD,GRU71HD,GRU81HD,FACTOR07
0,01,010101,Sierra Norte,"De 20,000 a 49,999 habitantes",No Pobre,No Vulnerable,52162.609375,34188.218750,2,3845.363770,874.391602,1006.782715,1307.083496,3339.000000,3247.260742,1318.787598,3219.265869,79.816757
1,01,010101,Sierra Norte,"De 20,000 a 49,999 habitantes",No Pobre,No Vulnerable,40832.042969,40164.945312,3,12792.489258,2384.777832,916.000000,665.926880,2408.242676,2768.000000,895.084717,752.953064,79.816757
2,01,010101,Sierra Norte,"De 20,000 a 49,999 habitantes",No Pobre,No Vulnerable,15098.497070,12308.838867,1,2150.485107,603.583740,2589.000000,581.172546,2047.000000,233.000000,41.478962,557.595886,79.816757
3,01,010101,Sierra Norte,"De 20,000 a 49,999 habitantes",No Pobre,No Vulnerable,41082.953125,30316.724609,2,9690.755859,39.434135,3178.366211,2735.143311,652.000000,930.000000,0.000000,1418.503906,79.816757
4,01,010101,Sierra Norte,"De 20,000 a 49,999 habitantes",No Pobre,Vulnerable No Pobre,47659.160156,33076.910156,5,5630.693359,1683.596191,1686.000000,797.020874,1372.000000,3086.147705,103.000000,894.525024,79.816757


## 3.1 Validación posterior al mapping

Se revisa que el mapeo de categorías no haya generado valores nulos y que los tipos de datos sean adecuados para el análisis.

In [7]:
print("Nulos después del mapping:")
print(df.isnull().sum()[df.isnull().sum() > 0])

print("\nTipos después del mapping:")
print(df.dtypes)

print("\nCardinalidad después del mapping:")
print(df.nunique(dropna=True))

Nulos después del mapping:
Series([], dtype: int64)

Tipos después del mapping:
MES             str
UBIGEO          str
DOMINIO         str
ESTRATO         str
POBREZA         str
POBREZAV        str
INGHOG2D    float64
GASHOG2D    float64
MIEPERHO      int64
GRU11HD     float64
GRU21HD     float64
GRU31HD     float64
GRU41HD     float64
GRU51HD     float64
GRU61HD     float64
GRU71HD     float64
GRU81HD     float64
FACTOR07    float64
dtype: object

Cardinalidad después del mapping:
MES            12
UBIGEO       1271
DOMINIO         8
ESTRATO         8
POBREZA         3
POBREZAV        4
INGHOG2D    33644
GASHOG2D    33675
MIEPERHO       17
GRU11HD     33330
GRU21HD     23069
GRU31HD      8769
GRU41HD     31258
GRU51HD      6744
GRU61HD     22348
GRU71HD     16887
GRU81HD     31407
FACTOR07      968
dtype: int64


## 4. Tratamiento de Outliers — Imputación con Mediana por Grupo

Se identifican valores atípicos en `INGHOG2D` y `GASHOG2D` mediante el método del Rango Intercuartílico (IQR).

**Decisión metodológica:** a diferencia de una eliminación de filas, se opta por **imputar los valores extremos con la mediana del grupo de pobreza** al que pertenece cada hogar. Esta decisión se fundamenta en que:

- ENAHO puede contener hogares con ingresos o gastos altos **reales y válidos**. Eliminarlos sesga el análisis de brechas al excluir precisamente los hogares más alejados de la línea de pobreza.
- Imputar con la mediana **del propio grupo** (`POBREZA`) preserva el perfil socioeconómico del hogar: un "No Pobre" con ingreso extremo recibe la mediana de los "No Pobres", no la mediana global.
- Se conservan las **33,691 filas originales**, garantizando representatividad estadística completa.

In [8]:
def imputar_outliers_mediana_grupo(dataset, columna, grupo):
    """
    Imputa valores atípicos (IQR) con la mediana del grupo al que pertenece cada hogar.
    No elimina filas: preserva el dataset completo.

    Parámetros
    ----------
    dataset : pd.DataFrame
    columna : str  — variable numérica a tratar
    grupo   : str  — variable categórica para calcular la mediana condicional
    """
    Q1 = dataset[columna].quantile(0.25)
    Q3 = dataset[columna].quantile(0.75)
    IQR = Q3 - Q1
    limite_inf = max(0, Q1 - 1.5 * IQR)
    limite_sup = Q3 + 1.5 * IQR

    outliers_mask = (dataset[columna] < limite_inf) | (dataset[columna] > limite_sup)
    medianas_grupo = dataset.groupby(grupo)[columna].median()

    dataset = dataset.copy()
    for idx in dataset[outliers_mask].index:
        cat = dataset.loc[idx, grupo]
        dataset.loc[idx, columna] = medianas_grupo[cat]

    print(f"Variable evaluada  : {columna}")
    print(f"Q1                 : {Q1:.2f}")
    print(f"Q3                 : {Q3:.2f}")
    print(f"IQR                : {IQR:.2f}")
    print(f"Límite inferior    : {limite_inf:.2f}")
    print(f"Límite superior    : {limite_sup:.2f}")
    print(f"Outliers imputados : {outliers_mask.sum()} ({outliers_mask.mean()*100:.2f}%)")
    print(f"Medianas por grupo usadas:")
    for cat, med in medianas_grupo.items():
        print(f"  {cat}: {med:.2f}")
    print()

    return dataset


df_limpio = df.copy()
print(f"Filas antes del tratamiento: {len(df_limpio)}\n")

# Imputar INGHOG2D (ingreso del hogar)
print("=" * 50)
df_limpio = imputar_outliers_mediana_grupo(df_limpio, "INGHOG2D", "POBREZA")

# Imputar GASHOG2D (gasto del hogar)
print("=" * 50)
df_limpio = imputar_outliers_mediana_grupo(df_limpio, "GASHOG2D", "POBREZA")

print(f"Filas después del tratamiento: {len(df_limpio)}")
print("No se eliminó ninguna fila. Todos los hogares conservados.")

Filas antes del tratamiento: 33691

Variable evaluada  : INGHOG2D
Q1                 : 15495.03
Q3                 : 47507.70
IQR                : 32012.67
Límite inferior    : 0.00
Límite superior    : 95526.70
Outliers imputados : 2037 (6.05%)
Medianas por grupo usadas:
  No Pobre: 31776.18
  Pobre Extremo: 10581.22
  Pobre No Extremo: 18182.92

Variable evaluada  : GASHOG2D
Q1                 : 14451.28
Q3                 : 37582.53
IQR                : 23131.25
Límite inferior    : 0.00
Límite superior    : 72279.40
Outliers imputados : 1589 (4.72%)
Medianas por grupo usadas:
  No Pobre: 27947.05
  Pobre Extremo: 8481.77
  Pobre No Extremo: 15380.45

Filas después del tratamiento: 33691
No se eliminó ninguna fila. Todos los hogares conservados.


## 4.2 Construcción de Variables Derivadas — Brecha Ingreso-Gasto

Esta sección construye las variables analíticas centrales para responder la pregunta del proyecto: **¿qué tan grande es la brecha entre ingresos y gastos en los hogares peruanos, y cómo varía según condición socioeconómica, territorio y composición del hogar?**

| Variable | Fórmula | Interpretación |
|---|---|---|
| `BRECHA_HOG` | `INGHOG2D - GASHOG2D` | Brecha absoluta del hogar. Positiva = ahorro, negativa = déficit. |
| `EN_DEFICIT` | `BRECHA_HOG < 0` | Flag: el hogar gasta más de lo que ingresa. |
| `ING_PERCAPITA` | `INGHOG2D / MIEPERHO` | Ingreso por miembro. Permite comparar hogares de distinto tamaño. |
| `GAS_PERCAPITA` | `GASHOG2D / MIEPERHO` | Gasto por miembro del hogar. |
| `BRECHA_PERCAPITA` | `BRECHA_HOG / MIEPERHO` | Brecha por miembro. Métrica comparable entre hogares. |

In [9]:
# --- Brecha absoluta del hogar ---
df_limpio['BRECHA_HOG'] = df_limpio['INGHOG2D'] - df_limpio['GASHOG2D']

# --- Flag de déficit ---
df_limpio['EN_DEFICIT'] = (df_limpio['BRECHA_HOG'] < 0).astype(int)

# --- Variables per cápita ---
df_limpio['ING_PERCAPITA']    = df_limpio['INGHOG2D']   / df_limpio['MIEPERHO']
df_limpio['GAS_PERCAPITA']    = df_limpio['GASHOG2D']   / df_limpio['MIEPERHO']
df_limpio['BRECHA_PERCAPITA'] = df_limpio['BRECHA_HOG'] / df_limpio['MIEPERHO']

# --- Resumen por condición de pobreza ---
print("Resumen de BRECHA_HOG por condición de pobreza:")
resumen_brecha = df_limpio.groupby('POBREZA').agg(
    Hogares        = ('BRECHA_HOG', 'count'),
    Brecha_Mediana = ('BRECHA_HOG', 'median'),
    Brecha_Media   = ('BRECHA_HOG', 'mean'),
    Pct_Deficit    = ('EN_DEFICIT', 'mean')
).round(2)
resumen_brecha['Pct_Deficit'] = (resumen_brecha['Pct_Deficit'] * 100).round(1)
print(resumen_brecha)

print(f"\nTotal hogares en déficit: {df_limpio['EN_DEFICIT'].sum()} ({df_limpio['EN_DEFICIT'].mean()*100:.1f}%)")

df_limpio[['POBREZA', 'INGHOG2D', 'GASHOG2D', 'MIEPERHO',
           'BRECHA_HOG', 'EN_DEFICIT', 'ING_PERCAPITA',
           'GAS_PERCAPITA', 'BRECHA_PERCAPITA']].head(8)

Resumen de BRECHA_HOG por condición de pobreza:
                  Hogares  Brecha_Mediana  Brecha_Media  Pct_Deficit
POBREZA                                                             
No Pobre            26879         2568.57       4828.21         38.0
Pobre Extremo        1588         1974.34       4598.89         32.0
Pobre No Extremo     5224         2399.17       5451.55         34.0

Total hogares en déficit: 12581 (37.3%)


,POBREZA,INGHOG2D,GASHOG2D,MIEPERHO,BRECHA_HOG,EN_DEFICIT,ING_PERCAPITA,GAS_PERCAPITA,BRECHA_PERCAPITA
0,No Pobre,52162.609375,34188.218750,2,17974.390625,0,26081.304688,17094.109375,8987.195312
1,No Pobre,40832.042969,40164.945312,3,667.097656,0,13610.680990,13388.315104,222.365885
2,No Pobre,15098.497070,12308.838867,1,2789.658203,0,15098.497070,12308.838867,2789.658203
3,No Pobre,41082.953125,30316.724609,2,10766.228516,0,20541.476562,15158.362305,5383.114258
4,No Pobre,47659.160156,33076.910156,5,14582.250000,0,9531.832031,6615.382031,2916.450000
5,No Pobre,37066.488281,31072.937500,4,5993.550781,0,9266.622070,7768.234375,1498.387695
6,No Pobre,28766.123047,37737.890625,3,-8971.767578,1,9588.707682,12579.296875,-2990.589193
7,No Pobre,41758.007812,29192.025391,2,12565.982422,0,20879.003906,14596.012695,6282.991211


## 4.3 Variables de Soporte para Visualización — Entrega 3

Esta sección construye variables adicionales que habilitan los cuatro tipos de vista requeridos en la Entrega 3. No modifican la información analítica del dataset: son transformaciones de soporte para Tableau.

| Variable nueva | Vista que habilita | Justificación técnica |
|---|---|---|
| `MES_NUM` | Tiempo | `MES` como string puede ordenarse lexicográficamente en lugar de cronológicamente en Tableau. La versión entera garantiza orden correcto en ejes temporales. |
| `DEPARTAMENTO` | Comparación geográfica | `UBIGEO` de 6 dígitos no tiene semántica legible. Se extrae el código de departamento (primeros 2 dígitos) y se mapea al nombre oficial INEI para comparaciones y mapas departamentales. |
| `LOG_INGHOG2D` / `LOG_GASHOG2D` / `LOG_BRECHA_PERCAP` | Distribución | Ingresos y gastos tienen distribución fuertemente sesgada a la derecha. La transformación logarítmica permite leer histogramas y box plots sin que la cola aplaste la varianza central. |
| `GRU{i}1HD_PCT` | Relación / composición | Los montos brutos de cada rubro son incomparables entre hogares de distinto nivel de ingreso. La proporción sobre el gasto total normaliza la comparación y es la entrada correcta para scatter plots y PCA. |

In [10]:
# ── Vista de TIEMPO ──────────────────────────────────────────────────────────
# MES como entero para ordenamiento correcto en ejes temporales de Tableau
df_limpio['MES_NUM'] = df_limpio['MES'].astype(int)

print("MES_NUM — valores únicos:", sorted(df_limpio['MES_NUM'].unique()))

MES_NUM — valores únicos: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12)]


In [11]:
# ── Vista de COMPARACIÓN GEOGRÁFICA ─────────────────────────────────────────
# Departamento a partir de los primeros 2 dígitos del UBIGEO (catálogo INEI)
dep_nombres = {
    '01': 'Amazonas',     '02': 'Áncash',        '03': 'Apurímac',
    '04': 'Arequipa',     '05': 'Ayacucho',      '06': 'Cajamarca',
    '07': 'Callao',       '08': 'Cusco',          '09': 'Huancavelica',
    '10': 'Huánuco',      '11': 'Ica',            '12': 'Junín',
    '13': 'La Libertad',  '14': 'Lambayeque',    '15': 'Lima',
    '16': 'Loreto',       '17': 'Madre de Dios',  '18': 'Moquegua',
    '19': 'Pasco',        '20': 'Piura',          '21': 'Puno',
    '22': 'San Martín',   '23': 'Tacna',          '24': 'Tumbes',
    '25': 'Ucayali'
}
df_limpio['DEPARTAMENTO'] = df_limpio['UBIGEO'].str[:2].map(dep_nombres)

print("DEPARTAMENTO — cardinalidad:", df_limpio['DEPARTAMENTO'].nunique())
print("Departamentos sin mapear  :", df_limpio['DEPARTAMENTO'].isnull().sum())
print()
print(df_limpio['DEPARTAMENTO'].value_counts())

DEPARTAMENTO — cardinalidad: 25
Departamentos sin mapear  : 0

DEPARTAMENTO
Lima             4544
Piura            1611
La Libertad      1509
Arequipa         1506
Junín            1506
Ica              1500
Cajamarca        1469
Áncash           1393
Loreto           1359
San Martín       1347
Lambayeque       1302
Tacna            1271
Huánuco          1244
Amazonas         1243
Cusco            1230
Ayacucho         1160
Puno             1123
Ucayali          1119
Huancavelica     1042
Callao           1027
Moquegua          963
Apurímac          913
Pasco             865
Tumbes            842
Madre de Dios     603
Name: count, dtype: int64


In [12]:
# ── Vista de DISTRIBUCIÓN ────────────────────────────────────────────────────
# Transformación logarítmica para variables con cola larga a la derecha.
# Se usa log1p (log(1+x)) para manejar correctamente valores en cero.
# BRECHA_PERCAPITA puede ser negativa (hogares en déficit), por lo que
# se aplica clip(lower=0) antes del log: los déficits se leen en la escala
# original; el log solo comprime la cola positiva (ahorro).
df_limpio['LOG_INGHOG2D']      = np.log1p(df_limpio['INGHOG2D'])
df_limpio['LOG_GASHOG2D']      = np.log1p(df_limpio['GASHOG2D'])
df_limpio['LOG_BRECHA_PERCAP'] = np.log1p(df_limpio['BRECHA_PERCAPITA'].clip(lower=0))

print("Estadísticas — INGHOG2D original vs LOG:")
print(df_limpio[['INGHOG2D', 'LOG_INGHOG2D']].describe().round(2))

print("\nEstadísticas — GASHOG2D original vs LOG:")
print(df_limpio[['GASHOG2D', 'LOG_GASHOG2D']].describe().round(2))

Estadísticas — INGHOG2D original vs LOG:
       INGHOG2D  LOG_INGHOG2D
count  33691.00      33691.00
mean   31106.98         10.11
std    20259.28          0.75
min       25.79          3.29
25%    15492.15          9.65
50%    27681.14         10.23
75%    41432.10         10.63
max    95516.21         11.47

Estadísticas — GASHOG2D original vs LOG:
       GASHOG2D  LOG_GASHOG2D
count  33691.00      33691.00
mean   26192.93          9.98
std    15188.54          0.66
min      825.15          6.72
25%    14451.28          9.58
50%    23941.33         10.08
75%    34461.88         10.45
max    72264.78         11.19


In [13]:
# ── Vista de RELACIÓN / COMPOSICIÓN ─────────────────────────────────────────
# Participación porcentual de cada rubro de gasto sobre el gasto total del hogar.
# Normaliza la comparación entre hogares de distinto nivel de ingreso.
# Es la entrada correcta para scatter plots de composición y para PCA/clustering.
grupos = [f'GRU{i}1HD' for i in range(1, 9)]
nombres_grupos = {
    'GRU11HD': 'Alimentos',
    'GRU21HD': 'Vestido y calzado',
    'GRU31HD': 'Vivienda y servicios',
    'GRU41HD': 'Muebles y enseres',
    'GRU51HD': 'Salud',
    'GRU61HD': 'Transporte y comunicaciones',
    'GRU71HD': 'Esparcimiento y cultura',
    'GRU81HD': 'Otros bienes y servicios'
}

for col in grupos:
    df_limpio[f'{col}_PCT'] = (
        df_limpio[col] / df_limpio['GASHOG2D'].replace(0, np.nan)
    ).fillna(0).round(4)

# Verificación: la suma de los PCT debe aproximarse a 1.0 en cada fila
cols_pct = [f'{c}_PCT' for c in grupos]
suma_pct = df_limpio[cols_pct].sum(axis=1)
print("Suma de participaciones por hogar (esperado ≈ 1.0):")
print(suma_pct.describe().round(4))

# Participación media por rubro según condición de pobreza
print("\nParticipación media de cada rubro según condición de pobreza:")
resumen_pct = df_limpio.groupby('POBREZA')[cols_pct].mean().round(3)
resumen_pct.columns = [nombres_grupos[c.replace('_PCT', '')] for c in cols_pct]
print(resumen_pct.T.to_string())

Suma de participaciones por hogar (esperado ≈ 1.0):
count    33691.0000
mean         0.6288
std          0.4043
min          0.0000
25%          0.4484
50%          0.5820
75%          0.7054
max          7.0078
dtype: float64

Participación media de cada rubro según condición de pobreza:
POBREZA                      No Pobre  Pobre Extremo  Pobre No Extremo
Alimentos                       0.278          0.294             0.330
Vestido y calzado               0.033          0.031             0.029
Vivienda y servicios            0.073          0.057             0.066
Muebles y enseres               0.043          0.045             0.037
Salud                           0.048          0.017             0.024
Transporte y comunicaciones     0.074          0.043             0.045
Esparcimiento y cultura         0.053          0.024             0.029
Otros bienes y servicios        0.041          0.022             0.026


## 5. Validación Final del Dataset Limpio

Se valida que el dataset limpio no tenga nulos, duplicados ni ambigüedades básicas de tipos antes de exportarlo para Tableau.

También se documenta la cardinalidad final de todas las variables, incluyendo las nuevas variables derivadas.

In [14]:
print("Filas y columnas del dataset limpio antes de dropna:", df_limpio.shape)

nulos = df_limpio.isnull().sum()
print("Valores nulos por columna:")
print(nulos[nulos > 0])

duplicados_limpio = df_limpio.duplicated().sum()
print("\nDuplicados en dataset limpio:", duplicados_limpio)

# Si hubiera nulos por alguna conversión fallida, se eliminan para evitar problemas en Tableau.
df_limpio = df_limpio.dropna().copy()

print("\nFilas y columnas del dataset limpio después de dropna:", df_limpio.shape)
print("Duplicados finales:", df_limpio.duplicated().sum())

validacion_limpio = pd.DataFrame({
    "variable": df_limpio.columns,
    "tipo_limpio": df_limpio.dtypes.astype(str).values,
    "nulos": df_limpio.isnull().sum().values,
    "pct_nulos": (df_limpio.isnull().mean() * 100).round(2).values,
    "cardinalidad": df_limpio.nunique(dropna=True).values
})

validacion_limpio

Filas y columnas del dataset limpio antes de dropna: (33691, 36)
Valores nulos por columna:
Series([], dtype: int64)

Duplicados en dataset limpio: 0

Filas y columnas del dataset limpio después de dropna: (33691, 36)
Duplicados finales: 0


,variable,tipo_limpio,nulos,pct_nulos,cardinalidad
0,MES,str,0,0.0,12
1,UBIGEO,str,0,0.0,1271
2,DOMINIO,str,0,0.0,8
3,ESTRATO,str,0,0.0,8
4,POBREZA,str,0,0.0,3
5,POBREZAV,str,0,0.0,4
6,INGHOG2D,float64,0,0.0,31609
7,GASHOG2D,float64,0,0.0,32087
8,MIEPERHO,int64,0,0.0,17
9,GRU11HD,float64,0,0.0,33330


## 5.1 Modelo Analítico para Tableau

Para esta entrega se utiliza una tabla plana a nivel de hogar. No se aplican joins ni relaciones adicionales porque el análisis inicial se concentra en una sola fuente: ENAHO Sumaria 2024.

**Dimensiones principales:**
- `MES`, `MES_NUM`, `UBIGEO`, `DEPARTAMENTO`, `DOMINIO`, `ESTRATO`, `POBREZA`, `POBREZAV`
- `EN_DEFICIT` (flag binario, usable como dimensión o filtro)

**Métricas base:**
- `INGHOG2D`, `GASHOG2D` — ingresos y gastos del hogar
- `BRECHA_HOG` — brecha absoluta ingreso-gasto ← métrica central del proyecto
- `ING_PERCAPITA`, `GAS_PERCAPITA`, `BRECHA_PERCAPITA` — versiones per cápita
- `MIEPERHO`, `FACTOR07`

**Métricas de distribución (transformadas):**
- `LOG_INGHOG2D`, `LOG_GASHOG2D`, `LOG_BRECHA_PERCAP` — para histogramas y box plots

**Métricas de composición del gasto:**
- `GRU11HD` a `GRU81HD` — montos brutos por rubro
- `GRU11HD_PCT` a `GRU81HD_PCT` — participación porcentual de cada rubro

**Validación de cardinalidad esperada:**
- `MES` / `MES_NUM`: 12 valores.
- `DOMINIO`: 8 categorías. `ESTRATO`: 8 categorías.
- `POBREZA`: 3 categorías. `POBREZAV`: 4 categorías.
- `DEPARTAMENTO`: 25 valores (uno por departamento del Perú).
- `EN_DEFICIT`: 2 valores (0 y 1).
- `UBIGEO`: texto, conservado para mapas de nivel distrital.

## 6. Exportación del Dataset Limpio

In [15]:
import os
import csv
os.makedirs('../Data/limpio', exist_ok=True)

# Forzar el zfill justo antes de exportar para evitar cualquier pérdida del tipo string
df_limpio['UBIGEO'] = df_limpio['UBIGEO'].astype(str).str.zfill(6)
df_limpio['MES'] = df_limpio['MES'].astype(str).str.zfill(2)

# Exportamos con codificación utf-8-sig (con BOM) para asegurar compatibilidad de tildes en Excel/Tableau
# Usamos QUOTE_NONNUMERIC para forzar comillas en columnas de tipo string (como UBIGEO)
df_limpio.to_csv('../Data/limpio/Sumaria-2024_limpio.csv', index=False, encoding='utf-8-sig', quoting=csv.QUOTE_NONNUMERIC)
print('Dataset exportado exitosamente a Data/limpio/Sumaria-2024_limpio.csv')
print(f'Dimensiones finales: {df_limpio.shape}')
print(f'\nColumnas exportadas ({len(df_limpio.columns)}):')
for i, col in enumerate(df_limpio.columns, 1):
    print(f'  {i:02d}. {col}')

Dataset exportado exitosamente a Data/limpio/Sumaria-2024_limpio.csv
Dimensiones finales: (33691, 36)

Columnas exportadas (36):
  01. MES
  02. UBIGEO
  03. DOMINIO
  04. ESTRATO
  05. POBREZA
  06. POBREZAV
  07. INGHOG2D
  08. GASHOG2D
  09. MIEPERHO
  10. GRU11HD
  11. GRU21HD
  12. GRU31HD
  13. GRU41HD
  14. GRU51HD
  15. GRU61HD
  16. GRU71HD
  17. GRU81HD
  18. FACTOR07
  19. BRECHA_HOG
  20. EN_DEFICIT
  21. ING_PERCAPITA
  22. GAS_PERCAPITA
  23. BRECHA_PERCAPITA
  24. MES_NUM
  25. DEPARTAMENTO
  26. LOG_INGHOG2D
  27. LOG_GASHOG2D
  28. LOG_BRECHA_PERCAP
  29. GRU11HD_PCT
  30. GRU21HD_PCT
  31. GRU31HD_PCT
  32. GRU41HD_PCT
  33. GRU51HD_PCT
  34. GRU61HD_PCT
  35. GRU71HD_PCT
  36. GRU81HD_PCT


## 7. Conclusión de Entrega 2

El dataset limpio queda preparado para su conexión en Tableau. La entrega cubre perfilado, selección de variables, revisión de nulos, duplicados, cardinalidad, homologación de categorías, tratamiento de outliers, construcción de variables derivadas, bitácora de transformaciones y exportación del archivo limpio.

Las principales decisiones metodológicas fueron:

- Trabajar a nivel de hogar.
- Conservar variables temporales, geográficas, socioeconómicas y de gasto.
- Preservar `MES` y `UBIGEO` como texto.
- Usar una tabla plana para Tableau.
- **Imputar outliers de `INGHOG2D` y `GASHOG2D` con la mediana del grupo de pobreza** en lugar de eliminar filas, conservando los 33,691 hogares originales y evitando sesgo en el análisis de brechas.
- **Construir `BRECHA_HOG`** como métrica central del proyecto, junto con sus versiones per cápita y el flag `EN_DEFICIT` para identificar hogares en déficit financiero.
- **Preparar variables de soporte para visualización** (Sección 4.3): `MES_NUM` para ejes temporales, `DEPARTAMENTO` para comparación geográfica, transformaciones logarítmicas para distribuciones sesgadas, y proporciones `GRU{i}1HD_PCT` para análisis de composición del gasto y PCA.